### PR TITLE
add testcase for #2195 (security delays)

### PR DESCRIPTION
**Pull request by taokayan**
#2195 add delay validation for linkauth/updateauth. add test cases for unlinkauth.

### CODE REVIEW

File: tests/chain_tests/delay_tests.cpp [hunk 1]
```diff
@@ -923,6 +923,16 @@ BOOST_AUTO_TEST_CASE( link_delay_link_change_test ) { try {
...
 
+   try {
+      chain.push_action(config::system_account_name, contracts::linkauth::get_name(), tester_account, fc::mutable_variant_object()
+      ("account", "tester")
+      ("code", "currency")
+      ("type", "transfer")
+      ("requirement", "second"),
+      30, 3);
+      BOOST_FAIL("linkauth should have a min delay set by previous linkauth");
+   } catch (...) { }
```
Comment (brianjohnson5972) [inline tests/chain_tests/delay_tests.cpp]: please add specific exception here and expected content (if it is used in more than one place).   We don't want to say this is passing if another exception is being thrown.

Comment (brianjohnson5972) [reply tests/chain_tests/delay_tests.cpp]: Use  BOOST_REQUIRE_EXCEPTION (with a predicate that identifies the specific exception content, if needed)

File: tests/chain_tests/delay_tests.cpp [hunk 2]
```diff
@@ -1024,6 +1034,202 @@ BOOST_AUTO_TEST_CASE( link_delay_link_change_test ) { try {
...
+   BOOST_REQUIRE_EQUAL(asset::from_string("0.0000 CUR"), liquid_balance);
+
+   try {
+      chain.push_action(config::system_account_name, contracts::unlinkauth::get_name(), tester_account, fc::mutable_variant_object()
+      ("account", "tester")
+      ("code", "currency")
+      ("type", "transfer"),
+      30, 7);
+      BOOST_FAIL("unlink should have a min delay set by previous linkauth");
+   } catch (...) { }
```
Comment (brianjohnson5972) [inline tests/chain_tests/delay_tests.cpp]: Please us BOOST_REQUIRE_EXCEPTION with specific exception details.

File: tests/chain_tests/delay_tests.cpp [hunk 3]
```diff
@@ -1444,6 +1650,18 @@ BOOST_AUTO_TEST_CASE( canceldelay_test ) { try {
...
+      chain.push_action(config::system_account_name, contracts::updateauth::get_name(), tester_account, fc::mutable_variant_object()
+            ("account", "tester")
+            ("permission", "first")
+            ("parent", "active")
+            ("data",  authority(chain.get_public_key(tester_account, "first")))
+            ("delay", 0),
+            30, 7
+      );
+      BOOST_FAIL("updateauth should have a min delay set by previous updateauth");
+   } catch (...) { }
```
Comment (brianjohnson5972) [inline tests/chain_tests/delay_tests.cpp]: Please use BOOST_REQUIRE_EXCEPTION with specific exception predicate.

File: tests/chain_tests/delay_tests.cpp [hunk 4]
```diff
@@ -1024,6 +1034,202 @@ BOOST_AUTO_TEST_CASE( link_delay_link_change_test ) { try {
...
+   BOOST_REQUIRE_EQUAL(asset::from_string("0.0000 CUR"), liquid_balance);
+
+   chain.produce_blocks(16);
+
+   liquid_balance = get_currency_balance(chain, N(tester));
+   BOOST_REQUIRE_EQUAL(asset::from_string("100.0000 CUR"), liquid_balance);
+   liquid_balance = get_currency_balance(chain, N(tester2));
+   BOOST_REQUIRE_EQUAL(asset::from_string("0.0000 CUR"), liquid_balance);
+
+   // this transaction will be delayed 20 blocks
```
Comment (brianjohnson5972) [inline tests/chain_tests/delay_tests.cpp]: What does the second delayed transfer do for this test?  If it doesn't add to test coverage and just makes the test run longer, we should remove it.

Comment (brianjohnson5972) [reply tests/chain_tests/delay_tests.cpp]: (I am basing my statement off of the exception you are expecting for the unlink above)

**Comment by taokayan** [comment]
hi I've created another pull request replacing "try" into BOOST_REQUIRE_EXCEPTION. Please would you review it?

File: tests/chain_tests/delay_tests.cpp [hunk 5]
```diff
@@ -923,15 +923,19 @@ BOOST_AUTO_TEST_CASE( link_delay_link_change_test ) { try {
...
       ("code", "currency")
       ("type", "transfer")
       ("requirement", "second"),
-      30, 3);
-      BOOST_FAIL("linkauth should have a min delay set by previous linkauth");
-   } catch (...) { }
+      30, 3),
+      transaction_exception,
+      [] (const transaction_exception &e)->bool {
+         BOOST_REQUIRE_EQUAL(std::string("transaction validation exception"), e.what());
```
Comment (brianjohnson5972) [inline tests/chain_tests/delay_tests.cpp]: Please use  e.to_detail_string() and check for something like "greater than the delay specified" so we know that it is this exception, same for the other BOOST_REQUIRE_EXCEPTIONs that are similar to this.

**Pull request merged**